File: dart/lib/src/freja.dart
import 'age_verification.dart';
import 'exceptions.dart';
import 'poll_options.dart';
import 'transport.dart';

class FrejaAuthRequest {
  const FrejaAuthRequest({
    this.ssn,
    this.callbackUrl,
    this.minRegistrationLevel,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String? ssn;
  final String? callbackUrl;
  final String? minRegistrationLevel;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        if (ssn != null) 'ssn': ssn,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (minRegistrationLevel != null) 'minRegistrationLevel': minRegistrationLevel,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class FrejaBackchannelAuthRequest {
  const FrejaBackchannelAuthRequest({
    required this.ssn,
    required this.country,
    this.callbackUrl,
    this.minRegistrationLevel,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String ssn;
  final String country;
  final String? callbackUrl;
  final String? minRegistrationLevel;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        'ssn': ssn,
        'country': country,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (minRegistrationLevel != null) 'minRegistrationLevel': minRegistrationLevel,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class FrejaSignRequest {
  const FrejaSignRequest({
    required this.text,
    this.ssn,
    this.callbackUrl,
    this.minRegistrationLevel,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String text;
  final String? ssn;
  final String? callbackUrl;
  final String? minRegistrationLevel;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        'text': text,
        if (ssn != null) 'ssn': ssn,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (minRegistrationLevel != null) 'minRegistrationLevel': minRegistrationLevel,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class FrejaBackchannelSignRequest {
  const FrejaBackchannelSignRequest({
    required this.ssn,
    required this.country,
    required this.text,
    this.callbackUrl,
    this.minRegistrationLevel,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String ssn;
  final String country;
  final String text;
  final String? callbackUrl;
  final String? minRegistrationLevel;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        'ssn': ssn,
        'country': country,
        'text': text,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (minRegistrationLevel != null) 'minRegistrationLevel': minRegistrationLevel,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

sealed class FrejaStatus {
  factory FrejaStatus.fromJson(Map<String, dynamic> json) => switch (json['status']) {
        'PENDING' => FrejaPending.fromJson(json),
        'COMPLETED' => FrejaCompleted.fromJson(json),
        'FAILED' => FrejaFailed.fromJson(json),
        _ => throw FormatException('Unknown freja status: ${json['status']}'),
      };
}

final class FrejaPending implements FrejaStatus {
  const FrejaPending({
    required this.id,
    this.refId,
    required this.autoStartToken,
    required this.qrData,
  });

  final String id;
  final String? refId;
  final String autoStartToken;
  final String qrData;

  factory FrejaPending.fromJson(Map<String, dynamic> json) => FrejaPending(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        autoStartToken: json['autoStartToken'] as String,
        qrData: json['qrData'] as String,
      );
}

final class FrejaCompleted implements FrejaStatus {
  const FrejaCompleted({
    required this.id,
    this.refId,
    required this.ssn,
    required this.country,
    required this.name,
    required this.givenName,
    required this.surname,
    this.address,
    this.companySignatoryText,
  });

  final String id;
  final String? refId;
  final String ssn;
  final String country;
  final String name;
  final String givenName;
  final String surname;
  final String? address;
  final String? companySignatoryText;

  factory FrejaCompleted.fromJson(Map<String, dynamic> json) => FrejaCompleted(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        ssn: json['ssn'] as String,
        country: json['country'] as String,
        name: json['name'] as String,
        givenName: json['givenName'] as String,
        surname: json['surname'] as String,
        address: json['address'] as String?,
        companySignatoryText: json['companySignatoryText'] as String?,
      );
}

final class FrejaFailed implements FrejaStatus {
  const FrejaFailed({required this.id, this.refId, required this.error});

  final String id;
  final String? refId;
  final String error;

  factory FrejaFailed.fromJson(Map<String, dynamic> json) => FrejaFailed(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        error: json['error'] as String,
      );
}

class FrejaEndpoint {
  FrejaEndpoint(this._transport);
  final Transport _transport;

  Future<FrejaStatus> auth(FrejaAuthRequest req) async =>
      FrejaStatus.fromJson(await _transport.post('/v3/freja/auth', req.toJson()));
  Future<FrejaStatus> backchannelAuth(FrejaBackchannelAuthRequest req) async =>
      FrejaStatus.fromJson(await _transport.post('/v3/freja/backchannel/auth', req.toJson()));
  Future<FrejaStatus> sign(FrejaSignRequest req) async =>
      FrejaStatus.fromJson(await _transport.post('/v3/freja/sign', req.toJson()));
  Future<FrejaStatus> backchannelSign(FrejaBackchannelSignRequest req) async =>
      FrejaStatus.fromJson(await _transport.post('/v3/freja/backchannel/sign', req.toJson()));
  Future<FrejaStatus> authStatus(String id) async =>
      FrejaStatus.fromJson(await _transport.get('/v3/freja/auth/$id'));
  Future<FrejaStatus> signStatus(String id) async =>
      FrejaStatus.fromJson(await _transport.get('/v3/freja/sign/$id'));
  Future<void> cancelAuth(String id) => _transport.delete('/v3/freja/auth/$id');
  Future<void> cancelSign(String id) => _transport.delete('/v3/freja/sign/$id');
  Future<void> cancelAgeVerification(String id) => _transport.delete('/v3/freja/age-verification/$id');

  Future<FrejaStatus> waitForAuth(String id, {PollOptions opts = const PollOptions()}) =>
      _poll(() => authStatus(id), opts);
  Future<FrejaStatus> waitForSign(String id, {PollOptions opts = const PollOptions()}) =>
      _poll(() => signStatus(id), opts);

  Future<AgeVerificationStatus> ageVerification(AgeVerificationRequest req) async =>
      AgeVerificationStatus.fromJson(await _transport.post('/v3/freja/age-verification', req.toJson()));

  Future<AgeVerificationStatus> ageVerificationStatus(String id) async =>
      AgeVerificationStatus.fromJson(await _transport.get('/v3/freja/age-verification/$id'));

  Future<AgeVerificationStatus> waitForAgeVerification(String id, {PollOptions opts = const PollOptions()}) =>
      _pollAge(() => ageVerificationStatus(id), opts);

  Future<FrejaStatus> _poll(Future<FrejaStatus> Function() fn, PollOptions opts) async {
    final deadline = DateTime.now().add(opts.timeout);
    while (true) {
      final status = await fn();
      if (status is! FrejaPending) return status;
      if (DateTime.now().isAfter(deadline)) throw WaitException(timeout: true);
      await Future<void>.delayed(opts.interval);
    }
  }

  Future<AgeVerificationStatus> _pollAge(Future<AgeVerificationStatus> Function() fn, PollOptions opts) async {
    final deadline = DateTime.now().add(opts.timeout);
    while (true) {
      final status = await fn();
      if (status is! AgeVerificationPending) return status;
      if (DateTime.now().isAfter(deadline)) throw WaitException(timeout: true);
      await Future<void>.delayed(opts.interval);
    }
  }
}

File: swift/Sources/IdkollenClient/Freja.swift
import Foundation

// MARK: - Requests

public struct FrejaAuthRequest: Encodable {
    public let ssn: String?
    public let callbackUrl: String?
    public let minRegistrationLevel: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        ssn: String? = nil,
        callbackUrl: String? = nil,
        minRegistrationLevel: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.ssn = ssn
        self.callbackUrl = callbackUrl
        self.minRegistrationLevel = minRegistrationLevel
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct FrejaBackchannelAuthRequest: Encodable {
    public let ssn: String
    public let country: String
    public let callbackUrl: String?
    public let minRegistrationLevel: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        ssn: String,
        country: String,
        callbackUrl: String? = nil,
        minRegistrationLevel: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.ssn = ssn
        self.country = country
        self.callbackUrl = callbackUrl
        self.minRegistrationLevel = minRegistrationLevel
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct FrejaSignRequest: Encodable {
    public let text: String
    public let ssn: String?
    public let callbackUrl: String?
    public let minRegistrationLevel: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        text: String,
        ssn: String? = nil,
        callbackUrl: String? = nil,
        minRegistrationLevel: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.text = text
        self.ssn = ssn
        self.callbackUrl = callbackUrl
        self.minRegistrationLevel = minRegistrationLevel
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct FrejaBackchannelSignRequest: Encodable {
    public let ssn: String
    public let country: String
    public let text: String
    public let callbackUrl: String?
    public let minRegistrationLevel: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        ssn: String,
        country: String,
        text: String,
        callbackUrl: String? = nil,
        minRegistrationLevel: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.ssn = ssn
        self.country = country
        self.text = text
        self.callbackUrl = callbackUrl
        self.minRegistrationLevel = minRegistrationLevel
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

// MARK: - Status enum + variants

public enum FrejaStatus: Decodable {
    case pending(FrejaPending)
    case completed(FrejaCompleted)
    case failed(FrejaFailed)

    private enum DiscriminatorKey: String, CodingKey { case status }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DiscriminatorKey.self)
        let status = try container.decode(String.self, forKey: .status)
        switch status {
        case "PENDING":   self = .pending(try FrejaPending(from: decoder))
        case "COMPLETED": self = .completed(try FrejaCompleted(from: decoder))
        case "FAILED":    self = .failed(try FrejaFailed(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: .status, in: container, debugDescription: "Unknown freja status: \(status)")
        }
    }
}

public struct FrejaPending: Decodable {
    public let id: String
    public let refId: String?
    public let autoStartToken: String
    public let qrData: String
}

public struct FrejaCompleted: Decodable {
    public let id: String
    public let refId: String?
    public let ssn: String
    public let country: String
    public let name: String
    public let givenName: String
    public let surname: String
    public let address: String?
    public let companySignatoryText: String?
}

public struct FrejaFailed: Decodable {
    public let id: String
    public let refId: String?
    public let error: String
}

// MARK: - Endpoint

public final class FrejaEndpoint {
    private let transport: Transport
    internal init(_ transport: Transport) { self.transport = transport }

    public func auth(_ req: FrejaAuthRequest) async throws -> FrejaStatus {
        try await transport.post("/v3/freja/auth", body: req)
    }

    public func backchannelAuth(_ req: FrejaBackchannelAuthRequest) async throws -> FrejaStatus {
        try await transport.post("/v3/freja/backchannel/auth", body: req)
    }

    public func sign(_ req: FrejaSignRequest) async throws -> FrejaStatus {
        try await transport.post("/v3/freja/sign", body: req)
    }

    public func backchannelSign(_ req: FrejaBackchannelSignRequest) async throws -> FrejaStatus {
        try await transport.post("/v3/freja/backchannel/sign", body: req)
    }

    public func authStatus(id: String) async throws -> FrejaStatus {
        try await transport.get("/v3/freja/auth/\(id)")
    }

    public func signStatus(id: String) async throws -> FrejaStatus {
        try await transport.get("/v3/freja/sign/\(id)")
    }

    public func cancelAuth(id: String) async throws {
        try await transport.delete("/v3/freja/auth/\(id)")
    }

    public func cancelSign(id: String) async throws {
        try await transport.delete("/v3/freja/sign/\(id)")
    }

    public func waitForAuth(id: String, opts: PollOptions = .init()) async throws -> FrejaStatus {
        try await poll({ try await self.authStatus(id: id) }, opts: opts)
    }

    public func ageVerification(_ req: AgeVerificationRequest) async throws -> AgeVerificationStatus {
        try await transport.post("/v3/freja/age-verification", body: req)
    }

    public func ageVerificationStatus(id: String) async throws -> AgeVerificationStatus {
        try await transport.get("/v3/freja/age-verification/\(id)")
    }

    public func cancelAgeVerification(id: String) async throws {
        try await transport.delete("/v3/freja/age-verification/\(id)")
    }

    public func waitForAgeVerification(id: String, opts: PollOptions = .init()) async throws -> AgeVerificationStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await ageVerificationStatus(id: id)
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }

    public func waitForSign(id: String, opts: PollOptions = .init()) async throws -> FrejaStatus {
        try await poll({ try await self.signStatus(id: id) }, opts: opts)
    }

    private func poll(_ fn: @escaping () async throws -> FrejaStatus, opts: PollOptions) async throws -> FrejaStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await fn()
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }
}
